src/db.ts

```diff
@@ -0,0 +1,401 @@
+            .then((row) => {
+              resolve({
+                id: row.id,
+                organization: organization,
+                name: name,
+              });
+            })
+            .catch((err) => {
+              reject(err);
+            });
+        })
+        .catch((err) => {
+          reject(err);
+        });
+    });
+  }
+
+  /**
+   * Get a pending pull-request (PR) from the repo, repo's branch, and upstream repo.
+   * @param {Repo} repo the Repo object representing the forked repo.
+   * @param {Repo} upstreamRepo the Repo object representing the upstream repo.
+   * @param {string} branch the forked repo's branch.
+   * @return {Promise<PendingPR | null>} A promise that resolves to a PendingPR object or null when there isn't a
+   *   pending PR. It rejects to if the database query fails.
+   */
+  async getPendingPR(repo: Repo, upstreamRepo: Repo, branch: string): Promise<PendingPR | null> {
+    return new Promise((resolve, reject) => {
+      const sql = `
+                SELECT upstream_pr_ids, action, pr_id, github_issue
+                FROM pending_prs
+                WHERE repo_id=? AND upstream_repo_id=? AND branch=?
+            `;
+      this.get(sql, [repo.id, upstreamRepo.id, branch])
+        .then((result) => {
+          if (!result) {
+            resolve(null);
+            return;
+          }
+
+          const upstreamPRIDs = [];
+          for (const idStr of (result.upstream_pr_ids as string).split(",")) {
+            upstreamPRIDs.push(parseInt(idStr));
+          }
+
+          resolve({
+            repo: repo,
+            branch: branch,
+            upstreamRepo: upstreamRepo,
+            upstreamPRIDs: upstreamPRIDs,
+            action: result.action,
+            prID: result.pr_id,
+            githubIssue: result.github_issue,
+          });
+        })
+        .catch((err) => {
+          reject(err);
+        });
+    });
+  }
+
+  /**
+   * Initialize the database by enabling foreign keys and creating all necessary tables and indices if not present.
+   * @return {Promise<void>} a promise that resolves to nothing and rejects if the database couldn't be initialized.
+   */
+  async init(): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.db = new sqlite3.Database(this.dbPath, (err) => {
+        if (err) {
+          reject(err);
+          return;
+        }
+
+        const statements = ["PRAGMA foreign_keys = ON"].concat(createSQLStatements);
+        const doNextStatement = (index: number = 0) => {
+          this.run(statements[index])
+            .then(() => {
+              // On the last statement, resolve the outer promise.
+              if (index === statements.length - 1) {
+                resolve();
+                return;
+              }
+
+              doNextStatement((index += 1));
+            })
+            .catch((err) => {
+              reject(err);
+            });
+        };
+
+        doNextStatement();
```

Comment:
What's the good reason for this being recursive? Is there no good way to handle the rejections in a more normal loop?